app/src/main/java/com/android/bookswap/model/map/Geolocation.kt

```diff
@@ -0,0 +1,114 @@
+ * - To stop receiving updates, call stopLocationUpdates().
+ * - The current user's location can be checked using the latitude and longitude variables.
+ * - The current state of location updates can be checked using the isRunning variable.
+ *
+ * This class requires appropriate location permissions to function, including both foreground and
+ * optionally background location access.
+ */
+class Geolocation(private val activity: Activity) {
+
+  private val REQUEST_LOCATION_PERMISSION = 1
+  private val BACKGROUND_LOCATION_PERMISSION_REQUEST_CODE = 2
+  private val fusedLocationClient: FusedLocationProviderClient =
+      LocationServices.getFusedLocationProviderClient(activity)
+  val isRunning = mutableStateOf(false)
+  val latitude = mutableStateOf(Double.NaN) // Change to MutableState
+  val longitude = mutableStateOf(Double.NaN)
+
+  // Location request settings
+  private val locationRequest: LocationRequest =
+      LocationRequest.create().apply {
+        interval = 10000 // Update interval in milliseconds
+        fastestInterval = 5000 // Fastest update interval in milliseconds
+        priority = LocationRequest.PRIORITY_HIGH_ACCURACY
+      }
+
+  private val locationCallback =
+      object : LocationCallback() {
+        override fun onLocationResult(p0: LocationResult) {
+          p0.lastLocation?.let { location ->
+            // Handle the updated location here
+            latitude.value = location.latitude
+            longitude.value = location.longitude
+            // You can save this location or notify other parts of your app
+          }
+        }
+      }
+
+  // Request location permissions
+  private fun requestLocationPermissions() {
+    val permissions =
+        arrayOf(
+            Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION)
+    ActivityCompat.requestPermissions(activity, permissions, REQUEST_LOCATION_PERMISSION)
+  }
+
+  // Check if permissions are granted
+  private fun hasLocationPermissions(): Boolean {
+    return ActivityCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_FINE_LOCATION) ==
+        PackageManager.PERMISSION_GRANTED &&
+        ActivityCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_COARSE_LOCATION) ==
+            PackageManager.PERMISSION_GRANTED
+  }
+
+  @RequiresApi(Build.VERSION_CODES.Q)
+  private fun requestBackgroundPermissions() {
+    val permissions = arrayOf(Manifest.permission.ACCESS_BACKGROUND_LOCATION)
+    ActivityCompat.requestPermissions(
+        activity, permissions, BACKGROUND_LOCATION_PERMISSION_REQUEST_CODE)
+  }
+
+  @RequiresApi(Build.VERSION_CODES.Q)
+  private fun hasBackgroundPermissions(): Boolean {
+    return ActivityCompat.checkSelfPermission(
+        activity, Manifest.permission.ACCESS_BACKGROUND_LOCATION) ==
+        PackageManager.PERMISSION_GRANTED
+  }
+
+  // Start location updates
+  @SuppressLint("MissingPermission")
+  fun startLocationUpdates() {
```

Comment:
I don't know if it will be possible in the future, but there's a chance that startLocationUpdates might be called even when it’s already running. You should either check if it's running within this function or ensure the check is done before calling it